src/azure-cli/azure/cli/command_modules/vm/custom.py

```diff
@@ -660,11 +660,34 @@ def create_vm(cmd, vm_name, resource_group_name, image=None, size='Standard_DS1_
 
         nic_dependencies = []
         if vnet_type == 'new':
-            vnet_name = vnet_name or '{}VNET'.format(vm_name)
             subnet = subnet or '{}Subnet'.format(vm_name)
-            nic_dependencies.append('Microsoft.Network/virtualNetworks/{}'.format(vnet_name))
-            master_template.add_resource(build_vnet_resource(
-                cmd, vnet_name, location, tags, vnet_address_prefix, subnet, subnet_address_prefix))
+            vnet_exists = False
+            if vnet_name:
+                from azure.cli.command_modules.vm._vm_utils import check_existence
+                vnet_exists = \
+                    check_existence(cmd.cli_ctx, vnet_name, resource_group_name, 'Microsoft.Network', 'virtualNetworks')
+                if vnet_exists:
+                    from azure.cli.core.commands import cached_get, cached_put, upsert_to_collection
+                    from azure.cli.command_modules.vm._validators import get_network_client
+                    client = get_network_client(cmd.cli_ctx).virtual_networks
+                    vnet = cached_get(cmd, client.get, resource_group_name, vnet_name)
+
+                    Subnet = cmd.get_models('Subnet', resource_type=ResourceType.MGMT_NETWORK)
+                    subnet_obj = Subnet(
+                        name=subnet,
+                        address_prefixes=[subnet_address_prefix],
+                        address_prefix=subnet_address_prefix
+                    )
```

Comment:
Adding a subnet is not as simple as adding a new resource. We need to take the **vnet's address space** and **subnet's prefix** into consideration. 

It is very hard to deduce the **new subnet's prefix** based on the **existing vnet's address space** and **existing subnet's prefix**. Even Azure Portal doesn't do the decision for the user. When creating a subnet, it just leave the field `Address range (CIDR block)` empty.

Some example edge cases:

1. If the existing vnet has address space `10.3.0.0/16`, this command will try to create a subnet with default prefix `10.0.0.0/24` and will fail, because the prefix is not in the allowed address space:
    > Azure Error: NetcfgInvalidSubnet
    > Message: Subnet 'vm3Subnet' is not valid in virtual network 'vnet3'.

2. If the existing vnet's address space is the same as the existing subnet's prefix - both `10.0.0.0/16`, the subnet creation will fail in any case, because there is no space left

3. If the existing vnet's address space is `10.0.0.0/24`, how do we decide the mask of the subnet? 24, 25? It is not practical to do the decision for the user

I prefer to error out if vnet is specified while subnet is not. 

cc @MyronFanQiu

Reply:
Yes, we discussed the cases you mentioned above before. Previously, there are some problems with the vnet/subnet prefix. However, for this PR, we don't want to change these behaviors, just keep them as before. 

What we changed in this PR this that, when vnet is specified and exists, no matter subnet is specified or not, if the subnet not exists, create it without removing other existing subnets. 

As the cases above, if the user not specify subnet prefix, it will use a default value (10.0.0.0/24) which may be wrong, however this is not the problems we want to solve in the PR.

Reply:
to share some context, not related with this PR. There's default vnet/subnet azure CLI created when creating VM. This is convenient scenario CLI providing: minimum effort to get started on a VM.

Reply:
> 
> 
> Adding a subnet is not as simple as adding a new resource. We need to take the **vnet's address space** and **subnet's prefix** into consideration.
> 
> It is very hard to deduce the **new subnet's prefix** based on the **existing vnet's address space** and **existing subnet's prefix**.
> 
> This PR will always try to add subnet `10.0.0.0/24` to the vnet. It'll succeed only if `10.0.0.0/24` is valid within the vnet, which is very unlikely in most cases. In other cases:
> 
>     1. If the existing vnet `10.0.0.0/16` already has subnet `10.0.0.0/24`, the creation will fail because the space is already occupied
> 
>     2. If the existing vnet is `10.3.0.0/16`, the creation will fail because the prefix is not in the allowed address space:
>        > Azure Error: NetcfgInvalidSubnet
>        > Message: Subnet 'vm3Subnet' is not valid in virtual network 'vnet3'.
> 
>     3. If the existing vnet's address space is the same as the existing subnet's prefix - both `10.0.0.0/16`, the subnet creation will fail in any case, because there is no space left
> 
>     4. If the existing vnet's address space is `10.0.0.0/24`, how do we decide the mask of the subnet? 24, 25? It is not practical to do the decision for the user
> 
> 
> **My suggestion**: If `--vnet-name` and `--subnet` are neither specified, `az create` will detect if there is already a vnet and subnet in the resource group and put the vm there. We can apply the same when only `--vnet-name` is specified with an existing vnet - if there is a subnet, put the vm there; if not, error out, instead of calculating the available prefix and creating a subnet voluntarily.
> 
> cc @MyronFanQiu

1, retry 25, 26..., or 23, 22...
2, we can create subnet with same address of vnet
3. retry 25, 26..., or 23, 22...
4, Azure CLI always makes decision for users so that users can use a very simple command to create resources.

I agree it's complex and tricky. Users might be astonished when a strange subnet is created.